bofire/data_models/features/categorical.py

```diff
@@ -360,37 +359,43 @@ class CategoricalOutput(Output):
     order_id: ClassVar[int] = 8
 
     categories: TCategoryVals
-    objective: Annotated[
-        List[Annotated[float, Field(type=float, ge=0, le=1)]], Field(min_items=2)
-    ]
+    objective: CategoricalObjective = Field(
+        default_factory=lambda: CategoricalObjective(w=1.0)
+    )
 
     @validator("categories")
     def validate_categories_unique(cls, categories):
         """validates that categories have unique names
 
         Args:
-            categories (List[str]): List of category names
+            categories (Union[List[str], Tuple[str]]): List or tuple of category names
 
         Raises:
             ValueError: when categories have non-unique names
 
         Returns:
-            List[str]: List of the categories
+            Tuple[str]: Tuple of the categories
         """
-        categories = list(categories)
         if len(categories) != len(set(categories)):
             raise ValueError("categories must be unique")
-        return categories
+        return tuple(categories)
 
     @validator("objective")
     def validate_objective(cls, objective, values):
-        if len(objective) != len(values["categories"]):
+        weights = objective.weights
+        if len(weights) != len(values["categories"]):
             raise ValueError("Length of objectives and categories do not match.")
-        for o in objective:
-            if o > 1:
-                raise ValueError("Objective values has to be smaller equal than 1.")
-            if o < 0:
-                raise ValueError("Objective values has to be larger equal than zero")
+        for w in weights:
+            if w > 1:
+                raise ValueError("Objective weight has to be smaller equal than 1.")
+            if w < 0:
+                raise ValueError("Objective weight has to be larger equal than zero")
+        # Save the categories to the objective if they do not exist
+        objective.categories = (
```

Comment:
Hmm, why is this needed? This is not a good data structure from a software engineering point of view.

Reply:
Is this referring to the reassignment of the `objective.categories`? I can change this so that `CategoricalObjective` takes in both objective and categories arguments. Then the validator just makes sure the categories are the same. Does that seem better?

Initially, I thought the repetition of categories was redundant, but maybe it makes more sense?

Reply:
I meant that the `CategoricalObjective` should not hold the categories, it should only hold the `desirabilities` or `weights` and the categories should just be attributes of the `CategoricalOutput.`

Reply:
I see - I have updated accordingly

Reply:
Just coming back to this; is there a reason the `CategoricalObjective` should not hold the categories? I think there is an inherit challenge with decoupling the `CategoricalOutput` and `CategoricalObjective`, but both functionalities serve different purposes in BoFire...

It might make sense to have the objective hold both the categories and the desireabilities, that way the CategoricalOutput can inherit those values and do any text processing/whatever else is done with Outputs and the Objective can handle the actual computations